Review

adacameron opened the pull request.

adacameron committed:
Pet constructor function & test

adacameron committed:
set name property

adacameron committed:
added growUp method on Pet prototype

adacameron committed:
set hunger and fitness properties on Pet object; changed growUp method to increase hunger property by 5 and decrease fitness property by 3

adacameron committed:
added walk & feed functions

adacameron committed:
added haveBaby and adoptChild functions + tests

adacameron committed:
added adoptChild and haveBaby, refactored tests and pet.js

adacameron committed:
updated readme

adacameron committed:
readme style

adacameron committed:
added pet variable to tests

ruimiguelcorreia commented on .gitignore:
```diff
@@ -0,0 +1 @@
+node_modules
```
Someone is learning how to not push node_modules!

ruimiguelcorreia commented on README.md:
```diff
@@ -1,6 +1,20 @@
+ * Give your pet a check up:
+`pet.checkUp()`
+
+
```
I would, later on, once you have some free time, expand on this README. You can explain the approach and logic behind it! Companies tend to look at your github projects when you're looking for a first job.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+const Pet = require('../src/pet');
+
+describe('constructor', () => {
+  let pet;
```
You can defined pet, and the beforeEach() outside this describe() block, moving it to the global scope - which means you don't need to keep repeating a beforeEach statement per describe().

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('returns an object', () => {
+    expect(new Pet('Gerald')).toBeInstanceOf(Object);
```
Since you've defined pet, you can do a different assertion:
expect(pet).toBeInstanceOf(Pet);
pet - your instance; Pet - your prototype

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('increments the age by 1', () => {
+    pet.age = 0;
```
You don't need in this case because the age is 0 when pet is initialised.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+    pet.age = 0;
+    pet.growUp();
+
+    expect(pet.age).toEqual(1);
```
When dealing with primitive values, I would use .toBe() as a matcher. I would use .toEqual() when comparing complex ones, like arrays.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('increments the hunger by 5', () => {
+    pet.hunger = 0;
```
Same as above. It should be 0 initially, so no need to worry.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('decreases the fitness by 3', () => {
+    pet.fitness = 10;
```
Same as above.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('throws an error message if the pet is no longer alive', () => {
+    pet.age = 30;
```
In this case, I would set pet.fitness to 0, just cause you're testing the walk() function directly.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+  });
+
+  it('throws an error message if the pet is no longer alive', () => {
+    pet.age = 30;
```
Same as above but for hunger.

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+    pet.fitness = 0;
+    pet.checkUp();
+
+    expect(pet.checkUp()).toBe('I need a walk');
```
If fitness is 0, shouldn't the pet be dead?

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+describe('haveBaby', () => {
+  it('creates a child object as a property of the parent', () => {
+    let parent = new Pet('Dave');
+
```
You have a parent - pet.
All you need is a child!

ruimiguelcorreia commented on __tests__/pet.test.js:
```diff
@@ -0,0 +1,292 @@
+describe('adoptChild', () => {
+  it('adds a child object to an array, which is the parent.children property', () => {
+
+    let parent = new Pet('Henry');
```
Same as above, you can use pet as the parent.

ruimiguelcorreia commented on src/pet.js:
```diff
@@ -0,0 +1,94 @@
+
+Pet.prototype = {
+  get isAlive() {
+    return this.age < 30 && this.hunger < 10 && this.fitness >= 0;
```
I would define a couple of constants for these too: MAXIMUM_AGE, MAXIMUM_HUNGER and MINIMUM_FITNESS.

ruimiguelcorreia commented on src/pet.js:
```diff
@@ -0,0 +1,94 @@
+  this.hunger += 5;
+  this.fitness -= 3;
+};
+
```
Same as above I would define a couple of constants for this: AGE_INCREMENT, etc.

ruimiguelcorreia replied:
And then, once you define its  increments/decrements, you can replace all of those magic numbers below.

ruimiguelcorreia commented on src/pet.js:
```diff
@@ -0,0 +1,94 @@
+
+  this.children.push(child);
+};
+
```
I like that you took 2 different approaches for haveBaby and adoptChild!

ruimiguelcorreia reviewed (approved):
Brilliant work Ada!
Just minor refactoring and you're good to go!

If you have some free time before BE, I would suggest that you come up with an interface for this logic!

adacameron committed:
Refactor after code review

Defined constant variables (and used them to replace magic numbers)

MAXIMUM_AGE
MAXIMUM_HUNGER
MINIMUM_FITNESS
FITNESS_INCREMENT
FITNESS_DECREMENT
AGE_INCREMENT
HUNGER_INCREMENT
HUNGER_DECREMENT

Tests:

- Moved beforeEach() and pet variable to the global scope (to avoid repetition)

- Changed the constructor assertion for 'returns an object'

- Removed unnecessary statements from growUp() feed() and walk() functions e.g. pet.age = 0

- Changed matchers with primitive values to toBe()

- walk() function error message - changed pet.age to pet.fitness
	- (to test walk() function directly)

- feed() function error message - changed pet.age to pet.hunger
	- (to test feed() function directly)

- checkUp() function - removed pet.fitness = 0
	- because it should throw the isAlive() function error (not 'I need a walk')
	- it is tested in the isAlive() function)

- Removed parent variable, and used the pet as the parent instead
